extensions/amp-experiment/1.0/apply-experiment.js

```diff
@@ -0,0 +1,155 @@
+ * @return {!Promise<!Object<string, ?string>>} a promise of the original
+ *     param passed in
+ * @private
+ */
+export function applyExperimentToVariant(ampdoc, config, experimentToVariant) {
+  // Get all of our mutation records across all experiments
+  // That are being applied
+  /** {Array<JsonObject>} */
+  let mutationRecords = [];
+  for (const experimentName in experimentToVariant) {
+    const variantName = experimentToVariant[experimentName];
+    if (variantName) {
+      const variantObject = config[experimentName]['variants'][variantName];
+      mutationRecords = mutationRecords.concat(variantObject['mutations']);
+    }
+  }
+
+  // Assert the formats of the mutation record,
+  // find its respective elements,
+  // count the number of mutations that will applied
+  const mutationRecordsAndElements = [];
+  let totalMutations = 0;
+  mutationRecords.forEach(mutationRecord => {
+    assertMutationRecordFormat(mutationRecord);
+
+    // Select the elements from the mutation record
+    const elements = getElementsFromMutationRecordSelector(
+      ampdoc.win.document,
+      mutationRecord
+    );
+    totalMutations += elements.length;
+    mutationRecordsAndElements.push({
+      mutationRecord,
+      elements,
+    });
+  });
+
+  if (totalMutations > MAX_MUTATIONS) {
+    const numMutationsError =
+      'Max number of mutations for the total ' +
+      `applied experiments exceeded: ${totalMutations} > ` +
+      MAX_MUTATIONS;
+    user().error(TAG, numMutationsError);
+    throw new Error(numMutationsError);
+  }
+
+  // Create the mutations
+  const mutations = [];
+  mutationRecordsAndElements.forEach(mutationRecordAndElements => {
+    const {mutationRecord, elements} = mutationRecordAndElements;
+
+    let mutation = undefined;
+    if (mutationRecord['type'] === 'characterData') {
+      mutation = new CharacterDataMutation(mutationRecord, elements);
+    } else if (mutationRecord['type'] === 'attributes') {
+      if (mutationRecord['attributeName'] === 'style') {
+        mutation = new AttributeMutationDefaultStyle(mutationRecord, elements);
+      } else if (
+        mutationRecord['attributeName'] === 'href' ||
+        mutationRecord['attributeName'] === 'src'
+      ) {
+        mutation = new AttributeMutationDefaultUrl(mutationRecord, elements);
+      } else if (mutationRecord['attributeName'] === 'class') {
+        mutation = new AttributeMutationDefaultClass(mutationRecord, elements);
+      } else {
+        // Did not find a supported attributeName
+        throw new Error(
+          `Mutation ${JSON.stringify(
+            mutationRecord
+          )} has an unsupported attributeName.`
+        );
+      }
+    } else {
+      user().error(
+        TAG,
+        'childList mutations not supported in the current experiment state.'
+      );
+
+      // TODO: Allow for innerHTML mutations
+      // Therefore, return a noop mutation.
+      mutation = {
```

Comment:
If there's a mutation that we don't support, do you think if we can throw an error and do not apply any mutation at all?

Reply:
The reason I am doing this here is to preserve backwards compatibility 😢 Though, I can totally throw an error here. Just want to double check, we don't have any partners in production currently right @jeffjose ?

Reply:
We've already support 'href', 'class' and 'style'. Is there any use case I'm missing that will lead to breaking change?

Reply:
Also @jeffjose and I ran a query last time, looks like we have few live pages that are using the feature right now. So I wouldn't worry too much about it. 

Could you provide an example on what will be broken? That would be very helpful.